design/proposal-crossplane-v2.md

```diff
@@ -0,0 +1,948 @@
+[persistent volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/)
+and extended it to handle any kind of infrastructure.
+
+I believe two things make Crossplane awkward for composing applications:
+
+1. Crossplane XRs and MRs are cluster scoped, while most Kubernetes resources
+   are namespaced
+2. Crossplane [doesn’t support](https://github.com/crossplane/crossplane/issues/1730)
+   directly composing arbitrary (non-Crossplane) Kubernetes resources
+
+When you build a Crossplane XR that represents an application today, it looks
+like this:
+
+```mermaid
+flowchart LR
+  subgraph ns [my-namespace]
+    app["Application (Claim)"]
+    dply["Deployment"]
+    svc["Service"]
+  end
+
+  %% Cluster scoped stuff
+  xapp["XApplication (XR)"]
+  obj1["Object (MR)"]
+  obj2["Object (MR)"]
+  rds["RDSInstance (MR)"]
+
+  app --> xapp
+  xapp --> obj1
+  xapp --> obj2
+  xapp --> rds
+  obj1 --> dply
+  obj2 --> svc
+```
+
+The application developer creates an Application in their namespace. This is a
+namespaced claim. Creating the claim ultimately causes Crossplane to create a
+namespaced Deployment and Service, and a cluster scoped RDSInstance. These are
+the composed resources. Between the claim and composed resources there’s a
+cluster scoped XR. For the Deployment and Service there’s also cluster scoped
+provider-kubernetes Object MRs.
+
+What value does the creator of the Application get from the intermediary cluster
+scoped resources? Why is the RDSInstance MR cluster scoped while the Deployment
+and Service aren’t? They’re all components of the same application.
+
+The following architecture achieves the same goal, but is simpler and more
+intuitive:
+
+```mermaid
+flowchart LR
+  subgraph ns [my-namespace]
+    direction LR
+    xr["Application (XR)"]
+  	dply["Deployment"]
+    svc["Service"]
+    rds["RDSInstance (MR)"]
+  end
+
+  xr --> dply
+  xr --> svc
+  xr --> rds
+```
+
+The namespaced Application abstraction is directly composed of the three
+resources that make up the logical application - a Deployment, a Service, and an
+RDSInstance. They’re all composed in the same namespace as the abstraction.
+
+This document proposes a set of changes to Crossplane that enable this simpler
+architecture.
+
+## Goals
+
+The goal of this proposal is to:
+
+* Make Crossplane less opinionated and more broadly useful
+* Make Crossplane more intuitive and easier to use
+* Avoid breaking existing Crossplane users
+
+These goals are in service of making Crossplane a better tool for composing
+applications _and_ infrastructure.
+
+## Proposal
+
+I propose we make the following changes in Crossplane v2.
+
+In core I propose we:
+
+* Make composite resources namespaced by default
+* Support composing any Kubernetes resource
+* Remove support for claims
+* Move all Crossplane-specific XR fields to `spec.crossplane`
+* Remove all deprecated features - e.g. native patch and transform composition
+
```

Comment:
One feature I am missing in this list is "implicit deletion dependencies". With the upcomming `Usage` resource we can build dependencies explicitly but that requires extra labour. And while it may sometimes be necessary in many cases an implicit DAG for resource should be sufficient. Similar tools like https://github.com/kro-run/kro already implement it.

Reply:
FWIW it's possible to implement kro-style implicit deletion dependencies in a function. I've done it in a private function that I can't share.

I don't think we're going to bring anything more than that into scope for v2.

Reply:
I have thought about writing a `function-usage` which would be similar to `function-sequencer` in that it would take a list of dependencies using composition-resource-names and generate a set of `Usage` resources that implement the dependencies specified.  I have a python implementation that I need to test more, but this would be a way to support a DAG for `dependsOn` deletion handling.